kornia/filters/canny.py

```diff
@@ -0,0 +1,167 @@
+        eps (float): regularization number to avoid NaN during backprop. Default: 1e-6.
+
+    Return:
+        torch.Tensor: the canny edge gradient magnitudes map with shape :math:`(B,C,H,W)`.
+
+    Example:
+        >>> input = torch.rand(1, 3, 4, 4)
+        >>> output = canny(input)  # 1x3x4x4
+        >>> output.shape
+        torch.Size([1, 3, 4, 4])
+    """
+    if not isinstance(input, torch.Tensor):
+        raise TypeError("Input type is not a torch.Tensor. Got {}"
+                        .format(type(input)))
+
+    if not len(input.shape) == 4:
+        raise ValueError("Invalid input shape, we expect BxCxHxW. Got: {}"
+                         .format(input.shape))
+
+    device: torch.device = input.device
+    dtype: torch.dtype = input.dtype
+
+    # Gaussian filter
+    blurred: torch.Tensor = gaussian_blur2d(input, kernel_size, sigma)
+
+    # Compute the gradients
+    edges: torch.Tensor = spatial_gradient(blurred, normalized=False)
+
+    # Unpack the edges
+    gx: torch.Tensor = edges[:, :, 0]
+    gy: torch.Tensor = edges[:, :, 1]
+
+    # Compute gradient maginitude and angle
+    magnitude: torch.Tensor = torch.sqrt(gx * gx + gy * gy + eps)
+    angle: torch.Tensor = torch.atan2(gy, gx)
+
+    # Radians to Degrees
+    angle = rad2deg(angle)
+
+    # Round angle to the nearest 45 degree
+    angle = torch.round(angle / 45) * 45
+
+    # Non-maximal suppression
+    nms_kernels: torch.Tensor = get_canny_nms_kernel(device, dtype)
+    nms_magnitude: torch.Tensor = F.conv2d(magnitude, nms_kernels, padding=nms_kernels.shape[-1]//2)
+
+    # Get the indices for both directions
+    positive_idx: torch.Tensor = (angle / 45) % 8
+    positive_idx = positive_idx.long()
+
+    negative_idx: torch.Tensor = ((angle / 45) + 4) % 8
+    negative_idx = negative_idx.long()
+
+    # Apply the non-maximum suppresion to the different directions
+    channel_select_filtered_positive: torch.Tensor = torch.gather(nms_magnitude, 1, positive_idx)
+    channel_select_filtered_negative: torch.Tensor = torch.gather(nms_magnitude, 1, negative_idx)
+
+    channel_select_filtered: torch.Tensor = torch.stack([channel_select_filtered_positive,channel_select_filtered_negative], 1)
+
+    is_max: torch.Tensor = channel_select_filtered.min(dim=1)[0] > 0.0
+
+    magnitude[is_max==0] = 0.0
+
+    # Threshold & Hysteresis
+    if low_threshold is not None:
+        magnitude = F.threshold(magnitude, low_threshold, 0.0)
+
+        if high_threshold is not None:
+            low: torch.Tensor = magnitude > low_threshold
+            high: torch.Tensor = magnitude > high_threshold
+
+            edges: torch.Tensor = low * 0.5 + high * 0.5
+
+            if hysteresis:
+
+                edges_old: torch.Tensor = torch.zeros(edges.shape, device=edges.device, dtype=edges.dtype)
+                hysteresis_kernels: torch.Tensor = get_hysteresis_kernel(device, dtype)
+
+                while ((edges_old - edges).abs() != 0).any():
```

Comment:
Isn`t this while loop slows down everything?
No offence/haven't done tests, just asking

Reply:
Yes, I agree, and it would be nice to avoid it. This is linked to the following issue #866  that should provides a better tool to do this step.

Reply:
@lferraz did some tricks with vmap/pmap around this for CLAHE that might be applicable here too. We can leave this a future improvement. Add a TODO note on behalf of that